src/aws-codeartifact/repository.ts

```diff
@@ -0,0 +1,406 @@
+   * @param id - The identifier of the construct.
+   * @param repositoryArn - The ARN of the repository to import.
+   */
+  public static fromRepositoryArn(scope: Construct, id: string, repositoryArn: string): IRepository {
+    const repositoryResourceArnParts = Stack.of(scope).splitArn(repositoryArn, ArnFormat.SLASH_RESOURCE_NAME);
+    if (
+      repositoryResourceArnParts.resource !== 'repository' ||
+      repositoryResourceArnParts.account === '' ||
+      repositoryResourceArnParts.resourceName === undefined
+    ) {
+      throw new Error(`Expected a repository ARN, but got ${repositoryArn}`);
+    }
+    const repositoryNameParts = repositoryResourceArnParts.resourceName.split('/');
+    if (repositoryNameParts.length !== 2) {
+      throw new Error(
+        `Expected a repository ARN with a domain and repository name (arn:aws:codeartifact:region-ID:account-ID:repository/my_domain/my_repo), but got ${repositoryArn}`,
+      );
+    }
+    const domainName = repositoryNameParts[0];
+    const repositoryName = repositoryNameParts[1];
+
+    const domain = Domain.fromDomainArn(
+      scope,
+      'Domain',
+      Stack.of(scope).formatArn({
+        resource: 'domain',
+        service: 'codeartifact',
+        resourceName: domainName,
+      }),
+    );
+
+    return Repository.fromRepositoryAttributes(scope, id, {
+      repositoryArn: repositoryArn,
+      repositoryName: repositoryName,
+      domain,
+    });
+  }
+
+  /**
+   * (internal) The CloudFormation resource representing this CodeArtifact repository.
+   */
+  protected cfnResource: CfnRepository;
+  /**
+   * The properties used to create the CloudFormation resource for this repository.
+   */
+  protected cfnResourceProps: CfnRepositoryProps;
+
+  /**
+   * TagManager to set, remove and format tags
+   */
+  readonly cdkTagManager: TagManager;
+  /**
+   * The ARN (Amazon Resource Name) of this CodeArtifact repository.
+   */
+  readonly repositoryArn: string;
+  /**
+   * The name of this CodeArtifact repository.
+   */
+  readonly repositoryName: string;
+  /**
+   * The domain that contains this repository.
+   */
+  readonly domain: IDomain;
+  protected readonly policy: PolicyDocument;
+
+  constructor(scope: Construct, id: string, props: RepositoryProps) {
+    super(scope, id);
+
+    this.cdkTagManager = new TagManager(TagType.KEY_VALUE, 'AWS::CodeArtifact::Repository');
+    this.policy = new PolicyDocument();
+
+    this.cfnResourceProps = {
+      domainName: props.domain.domainName,
+      repositoryName: props.repositoryName,
+      description: props.description,
+      domainOwner: props.domain.domainOwner,
+      externalConnections: props.externalConnection !== undefined ? [props.externalConnection] : undefined, // only 1 allowed
+      permissionsPolicyDocument: Lazy.any({ produce: () => this.policy?.toJSON() }),
+      upstreams: props.upstreams?.map(repo => repo.repositoryName),
```

Comment:
Cann we make that Lazy so we can add upstream via addUpstream methods?

Reply:
Made it lazy. Please have a look.